From a9e84a6d9a6e4c36f84543b1aa5303f99bca9b7a Mon Sep 17 00:00:00 2001
From: Timo Rothenpieler <timo@rothenpieler.org>
Date: Tue, 7 Jan 2025 19:17:54 +0100
Subject: [PATCH 14/21] UPSTREAM: avformat/flvdec: don't leak extradata pointer
 on realloc failure

(cherry picked from commit af74fe7139f4e99c12e9396b0b45c6d0c8d291cc)
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 libavformat/flvdec.c | 22 +++++++++++++---------
 1 file changed, 13 insertions(+), 9 deletions(-)

diff --git a/libavformat/flvdec.c b/libavformat/flvdec.c
index 88038332ea44..f167157a0463 100644
--- a/libavformat/flvdec.c
+++ b/libavformat/flvdec.c
@@ -973,18 +973,22 @@ static int flv_queue_extradata(FLVContext *flv, AVIOContext *pb, int stream,
         int new_count = stream + 1;
 
         if (flv->mt_extradata_cnt < new_count) {
-            flv->mt_extradata = av_realloc(flv->mt_extradata,
-                                           sizeof(*flv->mt_extradata) *
-                                           new_count);
-            flv->mt_extradata_sz = av_realloc(flv->mt_extradata_sz,
-                                              sizeof(*flv->mt_extradata_sz) *
-                                              new_count);
-            if (!flv->mt_extradata || !flv->mt_extradata_sz)
+            void *tmp = av_realloc_array(flv->mt_extradata, new_count,
+                                         sizeof(*flv->mt_extradata));
+            if (!tmp)
                 return AVERROR(ENOMEM);
+            flv->mt_extradata = tmp;
+
+            tmp = av_realloc_array(flv->mt_extradata_sz, new_count,
+                                   sizeof(*flv->mt_extradata_sz));
+            if (!tmp)
+                return AVERROR(ENOMEM);
+            flv->mt_extradata_sz = tmp;
+
             // Set newly allocated pointers/sizes to 0
             for (int i = flv->mt_extradata_cnt; i < new_count; i++) {
-                    flv->mt_extradata[i] = NULL;
-                    flv->mt_extradata_sz[i] = 0;
+                flv->mt_extradata[i] = NULL;
+                flv->mt_extradata_sz[i] = 0;
             }
             flv->mt_extradata_cnt = new_count;
         }
-- 
2.48.1.windows.1

